src/main/java/org/folio/circulation/domain/OverdueFineCalculatorService.java

```diff
@@ -55,37 +55,47 @@ private OverdueFineCalculatorService(Clients clients) {
       new OverduePeriodCalculatorService(new CalendarRepository(clients)));
   }
 
-  public CompletableFuture<Result<CheckInProcessRecords>> calculateOverdueFine(
+  public CompletableFuture<Result<LoanAndRelatedRecords>> calculateForRenewal(
+    LoanAndRelatedRecords records) {
+
+    return createOverdueFineIfNecessary(records, Scenario.RENEWAL);
+  }
+
+  public CompletableFuture<Result<CheckInProcessRecords>> calculateForCheckIn(
     CheckInProcessRecords records) {
 
-    Loan loan = records.getLoan();
+    return createOverdueFineIfNecessary(records, Scenario.CHECKIN);
+  }
+
+  <T extends LoanRelatedRecord> CompletableFuture<Result<T>> createOverdueFineIfNecessary(
+    T records, Scenario scenario) {
+
+    final Loan loan = records.getLoan();
 
-    if (loan == null) {
-      return completedFuture(of(() -> records));
+    if (loan == null || !loan.isOverdue()) {
+      return completedFuture(succeeded(records));
     }
 
-    return calculateOverdueFine(loan).thenApply(mapResult(v -> records));
+    return completedFuture(succeeded(loan))
+      .thenCompose(overdueFinePolicyRepository::findOverdueFinePolicyForLoan)
+      .thenCompose(r -> r.after(l -> createOverdueFineIfNecessary(l, scenario)))
+      .thenApply(mapResult(r -> records));
   }
 
-  public CompletableFuture<Result<LoanAndRelatedRecords>> calculateOverdueFine(
-    LoanAndRelatedRecords records) {
-
-    Loan loan = records.getLoan();
+  private CompletableFuture<Result<Loan>> createOverdueFineIfNecessary(Loan loan, Scenario scenario) {
+    CompletableFuture<Result<Loan>> loanFuture = completedFuture(succeeded(loan));
 
-    if (loan == null) {
-      return completedFuture(of(() -> records));
+    if (scenario.shouldNotCreateFine.test(loan.getOverdueFinePolicy())) {
+      return loanFuture;
     }
 
-    return calculateOverdueFine(loan).thenApply(mapResult(v -> records));
+    return loanFuture.thenCompose(r -> r.after(this::createOverdueFine));
   }
 
-  private CompletableFuture<Result<Account>> calculateOverdueFine(Loan loan) {
-    return completedFuture(succeeded(loan))
-      .thenComposeAsync(overdueFinePolicyRepository::findOverdueFinePolicyForLoan)
-      .thenCompose(r -> r.after(l -> getOverdueMinutes(l)
-          .thenCompose(mr -> mr.after(minutes -> calculateOverdueFine(l, minutes)))
-          .thenCompose(fr -> fr.after(fine -> this.createFeeFineRecord(l, fine)))
-      ));
+  private CompletableFuture<Result<Loan>> createOverdueFine(Loan loan) {
+    return getOverdueMinutes(loan)
+      .thenCompose(r -> r.after(minutes -> calculateOverdueFine(loan, minutes)))
+      .thenCompose(r -> r.after(fine -> createFeeFineRecord(loan, fine)));
```

Comment:
Why does creating a fine return a loan?

Reply:
Since we never use the result of creating a fine, and public methods of this class return their arguments, this particular method can return anything we want. But you are right, returning a `Loan` here does not make much sense. Replaced with `Void`.